### PR TITLE
Resuming large file uploads

### PR DESCRIPTION
**Pull request by bwbeach**
Replacing #139, this branch now includes unit tests.

### CODE REVIEW

File: test/test_bucket.py [hunk 1]
```diff
@@ -276,6 +277,95 @@ def test_upload_large(self):
...
+        large_file_id = self._start_large_file('file1', {'property': 'value1'})
+        self._upload_part(large_file_id, 1, data[:part_size])
+        progress_listener = StubProgressListener()
+        file_info = self.bucket.upload_bytes(
+            data,
+            'file1',
+            progress_listener=progress_listener,
+            file_infos={'property': 'value2'}
+        )
+        self.assertNotEqual(large_file_id, file_info.id_)
```
Comment (ppolewicz) [inline test/test_bucket.py]: please add `_check_file_contents` to check if the data which should have been uploaded, has been actually uploaded and has valid contents

File: test/test_bucket.py [hunk 2]
```diff
@@ -276,6 +277,95 @@ def test_upload_large(self):
...
+
+    def test_upload_large_resume_part_does_not_match(self):
+        part_size = self.simulator.MIN_PART_SIZE
+        data = self._make_data(part_size * 3)
+        large_file_id = self._start_large_file('file1')
+        self._upload_part(large_file_id, 3, data[:part_size])  # wrong part number for this data
+        progress_listener = StubProgressListener()
+        file_info = self.bucket.upload_bytes(data, 'file1', progress_listener=progress_listener)
+        self.assertNotEqual(large_file_id, file_info.id_)
+
```
Comment (ppolewicz) [inline test/test_bucket.py]: please add `_check_file_contents` to check if the data which should have been uploaded, has been actually uploaded and has valid contents

File: test/test_bucket.py [hunk 3]
```diff
@@ -276,6 +277,95 @@ def test_upload_large(self):
...
+        self._upload_part(large_file_id, 1, data[:part_size])
+        progress_listener = StubProgressListener()
+        file_info = self.bucket.upload_bytes(
+            data,
+            'file1',
+            progress_listener=progress_listener,
+            file_infos={'property': 'value2'}
+        )
+        self.assertNotEqual(large_file_id, file_info.id_)
+
```
Comment (ppolewicz) [inline test/test_bucket.py]: please add `_check_file_contents` to check if the data which should have been uploaded, has been actually uploaded and has valid contents

File: b2/bucket.py [hunk 4]
```diff
@@ -393,11 +400,62 @@ def _upload_large_file(
...
+                    if part_length != part.content_length:
+                        files_match = False
+                        break
+
+                    # Compare hash
+                    with upload_source.open() as f:
+                        f.seek(offset)
+                        sha1_sum = hex_sha1_of_stream(f, part_length)
+                    if sha1_sum != part.content_sha1:
+                        files_match = False
```
Comment (ppolewicz) [inline b2/bucket.py]: this branch is very important, but is not tested

**Comment by svonohr** [comment]
The SHA1 comparison is actually tested in `test_upload_large_resume_part_does_not_match`.

**Comment by ppolewicz** [comment]
Ah, right.

LGTM.

**Pull request merged**